Sender/tst/UT/BMSDataSender/testcases.c

```diff
@@ -0,0 +1,209 @@
+ * prepared as a hobby to save me from 
+ * - writing main
+ * - updating it with the newly added testcase
+ * - running the .exe after compiling my testcase (equivalent to run in actions)
+ * 
+ * The python script shall call this function before every testcase.
+ * All global variables are intialized to 0 in this function.
+ */ 
+
+static void Environment_Initialization(void)
+{
+    BMSDataTransmitter.TxControl.isTxStopRequested =0;
+    BMSDataTransmitter.TxControl.isStopAfterNTransmissionRequested =0;
+    BMSDataTransmitter.TxControl.NumberofTransmissionAllowed =0;
+    print = &printf;
+    Reset_all_print_mocks();
+    strcpy(filename,"./Sender/src/data.txt");
+}
+
+/**
+ * Sender will read the data from a file and transmit the data onto console.
+ * If the file is not found then the sender process shall exit.
+ */ 
+
+static void TC_ProcessShouldExitIfFileNotFound(void)
+{
+    strcpy(filename,"dat.txt");
+    printf("TC_ProcessShouldExitIfFileNotFound\n");
+    print = &print_Mock_FilenotfoundInstance;
+    BatteryMonitoringSystemTransmitter_Main();
+    assert(strcmp("Problem with File opening\n",Printf_FormartString)==0);
+}
+
+/**
+ * If the file present in the folder but unable to fetch it then
+ * testcase should fail.
+ * This checks if the file name is set to proper value or not
+ * This is a negative testcase
+ */  
+
+static void TC_EvaluateFilenameAndlocationofFile(void)
+{
+    printf("TC_EvaluateFilenameAndlocationofFile\n");
+    print = &print_Mock_FilenotfoundInstance;
+    BMSDataTransmitter.TxControl.isStopAfterNTransmissionRequested = 1;
+    BMSDataTransmitter.TxControl.NumberofTransmissionAllowed = 1;
+    BatteryMonitoringSystemTransmitter_Main();
+    assert(strcmp("Problem with File opening\n",Printf_FormartString)!=0);
+}
+
+
+
+
+/**
+ * The data is stored in the following order in the file
+ *      temparature_1 chargerate_1
+ *      temparature_2 chargerate_2
+ *      temparature_3 chargerate_3
+ *          ....
+ *          ....
+ * 
+ * The data should also get printed in the same order on console
+ *      temparature_1 chargerate_1
+ *      temparature_2 chargerate_2
+ *      temparature_3 chargerate_3
+ *          ....
+ *          ....
+ * This testcase will evaluate 
+ * 1. The way the data gets printed onto the console - "%f %f\n"
+ * 2. the first parameter that gets printed and second parameter that gets printed.
+ *    Temparature should be printed first
+ *    Chargerate should be printed second
+ */ 
+static void TC_EvaluateParametersOrderPrintedOnConsole(void)
+{
+    strcpy(filename,"./Sender/tst/UT/BMSDataSender/data.txt");
+    double constants[BatteryParameter_TotalNumber] = {1.0,2.0};
+    printf("TC_EvaluateParametersOrderPrintedOnConsole\n");
+    print = &print_Mock_ForDataEvaluation;
+    BatteryMonitoringSystemTransmitter_Main();
+    assert(call_Printf == 1);
+    assert(strcmp("%f %f\n",Printf_FormartString) == 0);
```

Comment:
Test cases need to reflect the functionality needed by the consumer. Drawbacks of this test are:
1. It's hard for someone to understand the output format, unless they read the comment. When parameters are added later, people may not update the comments
2. If the implementation doesn't use printf format strings, or restricts the number of decimals to two, is it a violation of user-need? May not be. Yet, the test will fail. The cost of writing this test is to make all implementation double work = implement in code and repeat it in the test